index: Force database compaction in coinstatsindex

fjahr opened the pull request:
This PR forces full database compaction on coinstatsindex every 10,000 blocks. This does not seem to come with any considerable performance impact, though I did not test on many different systems.

Motivation: Currently, coinstatsindex ends up using a lot of 55kb files if this is not the case. This is likely related to very fast writes occurring during sync but so far I could not pin down where exactly the difference in behavior is coming from. It seems to be a LevelDB internal thing though, since I can not make out any difference in configuration between coinstatsindex and the other indexes.

This was first reported here: https://github.com/bitcoin/bitcoin/pull/30469#issuecomment-3235597424

mzumsande reviewed (commented):
Do you know if the current state with many small files and one large file has actual downsides, or does it just look weird? 

I am not an expert on LevelDB, just mentioning that two AIs I consulted say that benefits of regular compaction would be only be of cosmetic value and suggest to just let LevelDB do its thing.

fjahr commented:
FWIW, from the patterns I have observed so far it seems likely to me that the problem is only occurring during sync and afterwards the data of new blocks are actually compacted as it should, just those 55kb files are left over from the sync phase. So I am thinking about changing the approach to only compact once when the index is being set to `m_synced = true`, though this would make it a more complex change and I need to do some more testing if my anecdotal observation is correct.

fjahr committed:
db, refactor: Add full compaction helper method

fjahr committed:
index: Force database compaction in coinstatsindex

The coinstatsindex ends up using a lot of 55kb files if this is not the case. This is likely related to very fast writes occurring during sync.

fjahr commented on src/index/coinstatsindex.cpp:
```diff
@@ -224,6 +224,14 @@ bool CoinStatsIndex::CustomAppend(const interfaces::BlockInfo& block)
+    // Force compaction of the database from time to time since LevelDB doesn't
+    // seem to be handling this well itself at the speed coinstats index is
+    // syncing.
+    if (block.height % 10000 == 0) {
```
Done

fjahr commented on src/index/coinstatsindex.cpp:
```diff
@@ -224,6 +224,14 @@ bool CoinStatsIndex::CustomAppend(const interfaces::BlockInfo& block)
+    // seem to be handling this well itself at the speed coinstats index is
+    // syncing.
+    if (block.height % 10000 == 0) {
+        LogInfo("Compacting database of coinstatsindex");
```
I am pretty indifferent, so making it debug. I had to add a new category since none of the existing seemed to fit.

fjahr commented on src/index/coinstatsindex.cpp:
```diff
@@ -224,6 +224,14 @@ bool CoinStatsIndex::CustomAppend(const interfaces::BlockInfo& block)
     m_muhash.Finalize(out);
     value.second.muhash = out;
 
+    // Force compaction of the database from time to time since LevelDB doesn't
+    // seem to be handling this well itself at the speed coinstats index is
+    // syncing.
```
As I and others have mentioned (see the linked PR), this doesn't work for coinstatsindex and we end up with hundreds of 55kb files that seem to stay around forever or until we trigger a full compaction manually.

Benchmarking didn't seem to make sense to me because the compaction completes almost infinitely so comparing two runs with compaction and without may not have given a result that gives an accurate answer. But I figured if I time the runtime of each compaction to the millisecond and then add them up that gives me a good indicator for how much time the compaction takes. I have been using Signet for this test and the individual runtimes of the compactions range from 12ms to 220ms. Overall the time for the coinstatsindex to sync (time between thread start and thread exit messages) took 5 minutes and 6 seconds. The cummulative time spend on compaction during that sync was 1,367ms. So the performance impact was making full sync ~0.45% slower.

Here is the patch I used for this: https://github.com/fjahr/bitcoin/commit/817990e515c1b64de59b2838c91032069651da5d In case you want to reproduce the result.

fjahr replied:
> could we rather add this to the other background compaction thread? Do we understand why there's no automatic compaction?

I am not aware of a compaction thread that we control, can you elaborate what you mean? And no, as I have tried to explain there seems to be no easy answer but the evidence points to the fact that LEVELDB can not keep up at a certain write speed.

> does it make sense to do the compaction so often? There's a reason that's usually not done for every change, garbage collection is usually a rare event. Wouldn't it suffice to to it once at the very end only? Maybe 10k isn't that often, I'll see if there's a way to measure it.

I had thought about doing it once at the end of sync and I am doing this now in the latest push. Just necessitates a hook to be added for this.

> Instead of the current single-entry LevelDB batching (creating a batch, serializing the DBHeightKey and std::pair<uint256, DBVal> and writing the batch, containing a single entry), I have extracted the CDBBatch outside of the loops and committing and writing that regularly,

Seems like a pretty big refactor and IMO not worth it only for the issue that we are dealing with here, given the edge cases and inconsistencies we have fixed in index over last few years.

l0rinc reviewed (changes requested):
I'm missing benchmark results to see how this change affects the already very slow performance. I will post my own results as soon as they finish.

I would also like to understand why database compaction isn't enabled for this LevelDB instance.

fjahr commented:
I am giving this another chance to receive some concept acks and see if the pain is big enough to add a fix.

Addressed comments by @l0rinc and particularly only compact once upon completion of sync. This required adding a hook for this to the base index but it's still a very minimal change.

There is still virtually zero performance impact from this change. I synced the `coinstatsindex` on signet on my machine and at some point there were 2500 of these 55kb files there, after compaction it was reduced to 4 files. That compaction process took less than a second (I didn't have more granular logging in place).
